rhn_register fixes for #24997 and #22300

WhyIsThisOpen opened the pull request:
##### SUMMARY

Require username and password for unregistering and avoid "cannot marshal None unless allow_none is enabled" error when using an activationkey and no channels specified.

Fixes for #24997 and #22300, including the changes from ansible/ansible@4f8dc60 and ansible/ansible@a4b69c1.

##### ISSUE TYPE

 - Bugfix Pull Request


##### COMPONENT NAME

    rhn_register.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
    ansible 2.3.0.0
      config file = /home/ansible/ansible.cfg
      configured module search path = Default w/o overrides
      python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```


##### ADDITIONAL INFORMATION
Undoes ansible/ansible@7faef2c.

WhyIsThisOpen committed:
rhn_register fixes for #24997 and #22300

Includes the changes from ansible/ansible@4f8dc60 and ansible/ansible@a4b69c1

bcoca commented:
cc @alikins, @mattclay

mmatuska commented:
Thanks, this makes Ansible 2.3 usable for us again!

mmatuska commented:
@bcoca could you please give a higher priority to this PR? It would be great to see it in 2.3.2 (rhn_register is broken for us since 2.3.0)

mmatuska commented:
bump @bcoca, @alikins, @mattclay

pilou- commented:
related: #26878

pilou- commented on lib/ansible/modules/packaging/os/rhn_register.py:
```diff
@@ -349,7 +351,13 @@ def main():
             channels=dict(default=[], type='list'),
-        )
+        ),
+        # username/password is required for state=absent, or if channels is not empty
```
when `state` is `present`: either (`username` and `password`) or `activation_key` are used in [`register`](https://github.com/WhyIsThisOpen/ansible/blob/57cf604488ed3051fa111a288f008fe6c3444d06/lib/ansible/modules/packaging/os/rhn_register.py#L252) method. `register` method is called independently from `channels` value.

Something like that could be used:
```
required_if=[
    ('state', 'present', ('username', 'activationkey'), True), # fourth item is True: at least one requirement should be present
    ('state', 'absent', ('username',)),
],
required_together=[('username', 'password')],

# then check values of channels and username, potentially calling 'module.fail_json'
```

See 61116a6f1fb56f784a77312e98c3f707577e649d.

maxamillion replied:
FWIW, this was discussed on IRC and @abadger @samdoran and myself agreed that while the `required_if` and `required_together` is more compact and a mechanism meant to be leveraged by modules, given the complexity of the decision matrix required it ends up being simpler and more maintainable code without using those.

alikins commented on lib/ansible/modules/packaging/os/rhn_register.py:
```diff
@@ -297,7 +297,9 @@ def unregister(self):
         os.unlink(self.config['systemIdPath'])
 
-    def subscribe(self, channels):
+    def subscribe(self, channels=[]):
```
Not sure I understand this. 

This makes channels a mutable default arg, meaning the same reference will be used for all calls of this method which is almost always the wrong behavior.

Why subscribe() need a default arg? it is only ever called from once place in module main, and it is called with the 'channels' var which is a list by default. How would subscribe() get called without channels being a list?

moshevayner replied:
@alikins I agree with you about the redundant initialization of channels in the subscribe method declaration. However at least in my case, the `if len(channels) <= 0:` statement at the beginning of this function is solving the bug mentioned here, in which the play would fail with the `cannot marshal None unless..` exception. Is this going to be included in the next release? 
Thanks!

mattclay replied:
If it's valid to call `subscribe` with an empty list or `None` for channels, use this instead:

```python
def subscribe(self, channels):
    if not channels:
         return
```

moshevayner replied:
@mattclay yep I agree, seems like the right way to do it!
So who can change the PR? I see it was opened by some kind of a bot? (WhyIsThisOpen).
I don't mind raising the glove and creating a new PR with the change you suggested, so that we'll have it included in the next release.

WhyIsThisOpen replied:
I am not a bot.

thaumos reviewed (dismissed):
Please do a rebase for this PR....

moshevayner commented:
Haha nice name selection then :-)
On Fri, Sep 1, 2017 at 3:50 PM WhyIsThisOpen <notifications@github.com>
wrote:

> *@WhyIsThisOpen* commented on this pull request.
> ------------------------------
>
> In lib/ansible/modules/packaging/os/rhn_register.py
> <https://github.com/ansible/ansible/pull/25079#discussion_r136652780>:
>
> > @@ -297,7 +297,9 @@ def unregister(self):
>          # Remove systemid file
>          os.unlink(self.config['systemIdPath'])
>
> -    def subscribe(self, channels):
> +    def subscribe(self, channels=[]):
>
> I am not a bot.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/25079#discussion_r136652780>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AL8b13bIU02Ld3oD24F11_OYkQHNI_s3ks5seGAbgaJpZM4Nn0ko>
> .
>
-- 

*Best Regards,*
*Moshe Shitrit.*

<https://www.linkedin.com/in/mosheshitrit>

angystardust commented:
Can @ansibot please add the label **affects_2.3** ?

angystardust commented:
@photoninger can you please create a fix for the 2.3 branch?

mpender commented:
Is this close to merging? would love to see it in 2.4

samdoran committed:
Do not make channels a mutable default arg

Leave the subscribe function if channels is the default value of [], which comes from the module argument spec.

samdoran commented:
Rebased and made suggested changes.

I don't have any way to test using this old method of registration. @MoShitrit would you be able to test this?

MoShitrit commented:
Hey @samdoran
Sorry but I won't be able to, I changed jobs and moved to another company recently, in which we're not using RHN Satellite anymore, so no way for me to test unfortunately.. 😐

angystardust commented:
@samdolan if you want, I should be able to test you patch against a Spacewalk 2.7 server

samdoran commented:
@MoShitrit No worries. Thanks for letting me know!

@angystardust If you could test and report back, I'd appreciate it. Thanks!

mattclay commented:
CI failure in unit tests:

```
>       assert mock_systemid.call_count == testcase['systemid.call_count']
E       AssertionError: assert 0 == 2
E        +  where 0 = <PropertyMock id='55563600'>.call_count

test/units/modules/packaging/os/test_rhn_register.py:254: AssertionError
```

samdoran committed:
Update fixtures and add changelog

angystardust commented:
@samdolan unfortunately the patch doesn't work on a 2.3 against a Spacewalk 2.7:

`fatal: [myremotehost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to 10.25.7.99 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_K74FPQ/ansible_module_rhn_register.py\", line 153, in <module>\r\n    from ansible.module_utils.six.moves import urllib, xmlrpc_client\r\nImportError: cannot import name xmlrpc_client\r\n", "msg": "MODULE FAILURE", "rc": 0}`

Stranglely the original patch from @WhyIsThisOpen works: https://raw.githubusercontent.com/WhyIsThisOpen/ansible/57cf604488ed3051fa111a288f008fe6c3444d06/lib/ansible/modules/packaging/os/rhn_register.py

Maybe it happen 'cause I'm on a 2.3 while the current patch works only for 2.4 setups?

samdoran commented:
@angystardust Yes, `six.moves` did not exist in 2.3, so it's failing the import. That doesn't test this fix.

maxamillion commented:
rebuild_merge

gundalow commented:
bot_status
